12-offer-encoding.md

```diff
@@ -0,0 +1,823 @@
+be uneven, with the deepest tree on the lowest-order leaves.
+
+e.g. consider the encoding of an `offer` `signature` with TLVs TLV1, TLV2 and TLV3:
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV1) 
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV2) 
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV3) 
+
+Assume L1 < L1nonce, L2 > L2nonce and L3 > L3nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightningoffersignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice: readers will either request an invoice
+(or multiple) or send an invoice based on the offer.  An offer can be much longer-lived than a
+particular invoice, so has some different characteristics; in particular it
+can be recurring, and the amount can be in a non-lightning currency.  It's
+also designed for compactness, to easily fit inside a QR code.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 6 (`currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 20 (`issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 22 (`quantity_min`)
+    2. data:
+        * [`tu64`:`min`]
+    1. type: 24 (`quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 30 (`node_id`)
+    2. data:
+        * [`point32`:`node_id`]
+    1. type: 54 (`send_invoice`)
+    1. type: 34 (`refund_for`)
+    2. data:
+        * [`sha256`:`refunded_payment_hash`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_path`
+2. data:
+   * [`point`:`blinding`]
+   * [`byte`:`num_hops`]
```

Comment:
Is this field necessary / desirable? Seems the number of hops can be determined from the TLV length (I think). And BOLT 1 recommends avoiding encoding the length twice where it can be avoided.

Reply:
`blinded_path` is not a TLV. The TLV contains a list of `blinded_path` so there is no way to derive the number of hops from the TLV length.

Reply:
Note that this doesn't decode with the latest draft at least (the path format changed.... @thomash-acinq is adding test vectors).